site-src/geps/gep-1762.md

```diff
@@ -0,0 +1,309 @@
+
+For example:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: default
+    port: 80
+    protocol: HTTP
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: merged-gateway
+spec:
+  infrastructure:
+    attachTo:
+      kind: Gateway
+      name: parent-gateway
+  gatewayClassName: example
+  listeners:
+  - name: some-other-route
+    port: 81
+    protocol: HTTP
+```
+
+With this configuration, an implementation:
+* MUST treat the `parent-gateway` as having the concatenated list of all listeners from merged Gateways.
+  This means, for example, that a Service with port 80 and 81 may be created in this example.
+
+This introduces a new concept in Gateway merging of a "primary" Gateway.
+This makes it explicit which is the primary gateway and which are attached to it.
+
+Secondary Gateways cannot configure `addresses` or anything in `infrastructure` (aside from `attachTo`), and must have
+the same `gatewayClassName` as the primary Gateway they are attached to.
+
+Gateways can only `attachTo` a primary Gateway. That is, we cannot have a chain of references.
+
+### Customizations
+
+With any in-cluster deployment, customization requirements will arise. 
+
+Some common requirements would be:
+* `Service.spec.type`, to control whether a service is a `ClusterIP` or `LoadBalancer`.
+* IP in the Service to assign to it.
+* Arbitrary labels and annotations on generated resources.
+* Any other arbitrary fields; the list is unbounded. Some examples would be:
+  * CPU and memory requests
+  * Service `externalTrafficPolicy`
+  * Affinity rules.
+
+#### Gateway Type
+
+This is handled by [GEP-1651](https://github.com/kubernetes-sigs/gateway-api/pull/1653), so won't be described here.
+
+#### Gateway IP
+
+For setting a specific IP address in a generated `Service`, the `.Spec.Address` field can be used.
+
+For example:
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: my-gateway
+spec:
+  addresses:
+  - type: IPAddress
+    value: 1.1.1.1
+  gatewayClassName: example
+  listeners:
+  - name: default
+    port: 80
+    protocol: HTTP
+```
+
+This would generate a `Service` with `clusterIP` or `loadBalancerIP`, depending on the Service type.
+
+Note: this differs from the selection of "manual deployment", which doesn't use `IPAddress` type.
+
+#### Labels and Annotations
+
+Labels and annotations SHOULD be copied from the `Gateway` to generated resources.
+
+Note this may mean an annotation intended for a `Service` may end up on a `Deployment` (for example).
+This is typically not a concern.
+However, if an implementation is aware of specific meanings of certain labels or annotations, they MAY
+exclude these from irrelevant resources.
+
+#### Arbitrary Customization
+
+Currently, to provide arbitrary customizations, Gateway API provides a few mechanisms:
```

Comment:
Not sure if 'arbitrary customization' should be in scope of the Gateway API. 

Since we support 'external' gateways - each implementation may have its own APIs and configs - in many cases outside of K8S. For in-cluster - if we allow users to create their own deployment and Service - they already have all they need to do arbitrary customization. 

I don't see a middle case where more arbitrary customizations would be needed - nor the use case for gateway API to prescribe how implementations handle their custom configs.

Reply:
Example use case: I want to set CPU requests to "2 CPUs". How can I do that?

If I create a deployment before I deploy the Gateway, I don't know what to put in there - there are 100s of fields like image, etc that I want the controller to apply for me.

If I patch the deployment afterwards, this has a few issues:
* Controller cannot do sophisticated merging logic. It cannot say "I want to default 1 CPU unless user has set it"; it can either force ownership or never set it, afaik
* We have some period of time with the wrong settings. For CPu its not so bad (besides needing to redeploy for no reason), but what if the user's customization was "do not run as root" for example

Both approaches are also tricky since they rely on ordering between actions, which is not very gitops/declarative friendly.

Reply:
I understand - my point was that it may be a broader problem ( anything that creates resources has it ), and may be handled by each implementation. 

For example in Istio we had a helm chart to install gateway without injection or controler involvment. Others may still do this, and it takes care of all options including image.

Having some template is also common - Deployment, Knative, etc have spec with a lot of options. Just not sure if we want to mix this broad problem into this specification. Would be a good thing to solve, but in a separate context ( and as for other proposals - after some survey on how different vendors deploy in cluster gateways ).